streamer/src/nonblocking/quic.rs

```diff
@@ -960,37 +958,40 @@ async fn packet_batch_sender(
                 Ok(packet_receiver.recv().await)
             };
 
-            if let Ok(Ok(packet_accumulator)) = timeout_res {
+            if let Ok(Ok(mut packet_accumulator)) = timeout_res {
                 // Start the timeout from when the packet batch first becomes non-empty
                 if packet_batch.is_empty() {
                     batch_start_time = Instant::now();
                 }
 
-                unsafe {
-                    let new_len = packet_batch.len() + 1;
-                    packet_batch.set_len(new_len);
-                }
-
-                let i = packet_batch.len() - 1;
-                *packet_batch[i].meta_mut() = packet_accumulator.meta;
+                // 86% of transactions/packets come in one chunk. In that case,
+                // we can just move the chunk to the `Packet` and no copy is
+                // made.
+                // 14% of them come in multiple chunks. In that case, we copy
+                // them into one `Bytes` buffer. We make a copy once, with
+                // intention to not do it again.
                 let num_chunks = packet_accumulator.chunks.len();
-                let mut offset = 0;
-                for chunk in packet_accumulator.chunks {
-                    packet_batch[i].buffer_mut()[offset..offset + chunk.len()]
-                        .copy_from_slice(&chunk);
-                    offset += chunk.len();
-                }
+                let mut packet = if packet_accumulator.chunks.len() == 1 {
+                    BytesPacket::new(
+                        packet_accumulator.chunks.pop().expect("expected one chunk"),
+                        packet_accumulator.meta,
+                    )
+                } else {
+                    let mut buf = BytesMut::with_capacity(PACKET_DATA_SIZE);
```

Comment:
instead of reserving PACKET_DATA_SIZE capacity, you should loop over the chunks
and get the exact capacity

Reply:
I understand this isn’t related to this PR, but if you guys are working on overall performance, initial deduplication can already be done right here. From my observations, 20–30% up to 70% (during periods of high activity) can be dropped at this level.

https://github.com/anza-xyz/agave/pull/4886#issuecomment-2798756858

Reply:
@alessandrod fixed

@diman-io let's talk in your PR